projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,171 @@
+  }
+  // if more than one channel is enabled go to next channel.
+  if (channel_enable != (1 << current_channel)) {
+    uint8_t temp = channel_enable;
+    temp &= ~((1 << (current_channel + 1)) - 1);
+    uint8_t index_ffs = __builtin_ffs(temp);
+    // The following if statement checks for the special case where
+    // the current channel is (NUM_ADS1015_CHANNELS - 1)
+    if (index_ffs) {
+      current_channel = index_ffs - 1;
+    } else {
+      current_channel = __builtin_ffs(channel_enable) - 1;
+    }
+    // Update so that the ADS1015 reads from the next channel.
+    prv_set_channel(storage, current_channel);
+  }
+}
+
+// Initiates ads1015 by setting up registers and enabling ALRT/RDY Pin.
+// It also registers the interrupt handler on ALRT/RDY pin.
+StatusCode ads1015_init(Ads1015Storage *storage, I2CPort i2c_port, Ads1015Address i2c_addr,
+                        GPIOAddress *ready_pin) {
+  if ((storage == NULL) || (ready_pin == NULL)) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  memset(storage, 0, sizeof(*storage));
+  storage->i2c_port = i2c_port;
+  storage->i2c_addr = i2c_addr;
+  storage->ready_pin = *ready_pin;
+  // Set up config register.
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_CONFIG,
+                                         ADS1015_CONFIG_REGISTER_MSB(ADS1015_CHANNEL_0),
+                                         ADS1015_CONFIG_REGISTER_LSB));
+  // Set up hi/lo-thresh registers. This setup enables the ALRT/RDY pin.
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_LO_THRESH,
+                                         ADS1015_LO_THRESH_REGISTER_MSB,
+                                         ADS1015_LO_THRESH_REGISTER_LSB));
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_HI_THRESH,
+                                         ADS1015_HI_THRESH_REGISTER_MSB,
+                                         ADS1015_HI_THRESH_REGISTER_LSB));
+  GPIOSettings gpio_settings = {
+    .direction = GPIO_DIR_IN,  //
+  };
+  InterruptSettings it_settings = {
+    .type = INTERRUPT_TYPE_INTERRUPT,
+    .priority = INTERRUPT_PRIORITY_NORMAL,
+  };
+  status_ok_or_return(gpio_init_pin(ready_pin, &gpio_settings));
+  status_ok_or_return(gpio_it_register_interrupt(ready_pin, &it_settings, INTERRUPT_EDGE_RISING,
+                                                 prv_interrupt_handler, storage));
+  // Mask the interrupt until channels are enabled by the user.
+  return gpio_it_mask_interrupt(ready_pin, true);
+}
+
+// This function enable/disables channels, and registers callbacks for each channel.
+StatusCode ads1015_configure_channel(Ads1015Storage *storage, Ads1015Channel channel, bool enable,
+                                     Ads1015Callback callback, void *context) {
+  if (storage == NULL || channel >= NUM_ADS1015_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  } else if (storage->channel_enable == 0 && enable) {
```

Comment:
Can you split the argument check from the logic?

I think it would also be more clear if you #defined "no channels" instead of just comparing it to 0 everywhere.